### PR TITLE
Apply guided decoding only for grammar requests

### PR DESCRIPTION
**Pull request by guoqingbao**
## What changed

- Extract only guided rows before applying guided decoding in mixed batches.
- Scatter the guided-logit delta back into the full F32 logits tensor and perform one sampling pass.
- Normalize logits to F32 once at the common `ModelRunner::sample()` boundary.
- Remove the redundant F32 conversion from `GuidedDecoding::apply()`.

## Why

The runner previously entered the guided-decoding path for every batch, including ordinary requests without tool grammar or structured-output constraints. In mixed batches, that also caused non-guided rows to pay the guidance masking cost.

The model audit found that standard safetensor paths already cast final logits to F32, while GGUF/QVar paths use QLinear, whose output is F32. The common sampling-boundary normalization protects graph and other model paths as well.

## Validation

- `cargo fmt -- --check`
- `git diff --check`
- `cargo build --release --features cuda`

### CODE REVIEW

**Comment by sempervictus** [comment]
Mind if I merge this into #421? I have structured reasoning with exclusion masks working ~95% (refining anchoring) which is highly complementary since using exclusion masks is very cheap relative to expressions but the operational cost you're skipping here ends up being the majority of the cost. With both of these PRs together we end up with an env var toggling between high security + optional advanced grammar features at the (IMO negligible) cost of guidance OR optimized industry standard behavior with minimal required coverage. Pairs really well IMO.

Ultimately I'm hoping to fuse the logic into the cuda sampling pass directly but that doesn't help with the setup cycles this is avoiding entirely

**Comment by guoqingbao** [comment]
> Mind if I merge this into #421? I have structured reasoning with exclusion masks working ~95% (refining anchoring) which is highly complementary since using exclusion masks is very cheap relative to expressions but the operational cost you're skipping here ends up being the majority of the cost. With both of these PRs together we end up with an env var toggling between high security + optional advanced grammar features at the (IMO negligible) cost of guidance OR optimized industry standard behavior with minimal required coverage. Pairs really well IMO.
> 
> Ultimately I'm hoping to fuse the logic into the cuda sampling pass directly but that doesn't help with the setup cycles this is avoiding entirely

I’ll merge this soon so you can rebase onto `main`.

**Comment by sempervictus** [comment]
Thank you!

Re the "reasoning guidance" piece, check this out:
image

^^ that's a qwen3.5 9B working out problem domain logic like some big fat minimax ... on a single V100 :grin: 

```rust
                    r#"start: reasoning_block
reasoning_block: preamble scientific_flow
preamble: "\nI will apply the scientific method with systematic observation, hypothesis testing, and evidence-based conclusion derivation to contextualize the concern:" text_inner
scientific_flow: observation_setup hypothesis_formulation test_design result_analysis conclusion_derivation
observation_setup: "\nFirst, I observe the following facts:" text_inner "\nThis raises the question:" text_inner
hypothesis_formulation: "\nBased on these observations, my working hypothesis is:" text_inner "\nIf true, we would expect to see:" text_inner
test_design: "\nTo test this hypothesis, I need to examine:" text_inner "\nThe key methodology involves:" text_inner
result_analysis: "\nUpon analysis, the data shows:" text_inner "\nThis either supports or refutes my hypothesis because:" text_inner
conclusion_derivation: "\nTherefore, based on this systematic analysis, I conclude:" text_inner "\nMy confidence in this conclusion is:" text_inner "\nThe evidence supporting this includes:" reasoning_text
text_inner[stop="\n\n", temperature=1.0]: /[\x20-\x7E\x0A\x0D]+?/
reasoning_text: {}
"#,
                    guidance_tokens.reasoning_grammar_mask();
```

Using exclusion masks is a huge win for stability since it gives the model a natural way to emit until it reaches the viable point for the next anchor. There's still _some_ instability in this approach with small models over multiple iterations (i have an issue up w/ LLG for this) and you have to _both_ enable `XINFER_LLG_FULL` _and_ send a `high`+ reasoning level request to the API as the client to trip that behavior but when its on, it works a charm.

**Comment by guoqingbao** [comment]
> ^^ that's a qwen3.5 9B working out problem domain logic like some big fat minimax ... on a single V100 😁

The 9B is a dense model which should works well than many MoE models, the 27B one is the best under 100B size, matching some of the 300B MoE models.

**Comment by sempervictus** [comment]
Point taken but what i'm hoping to illustrate is that you can make it traverse all those params in a very prescriptive way to figure out what it will actually do during the turn :-).
